src/backend/metal/src/device.rs

```diff
@@ -1354,43 +1354,48 @@ impl hal::Device<Backend> for Device {
 
             n::DescriptorSetLayout::ArgumentBuffer(encoder, stage_flags)
         } else {
+            struct TempSampler {
+                sampler: metal::SamplerState,
+                binding: pso::DescriptorBinding,
+                array_index: pso::DescriptorArrayIndex,
+            };
+            let mut immutable_sampler_iter = immutable_samplers.into_iter();
+            let mut tmp_samplers = Vec::new();
             let mut desc_layouts = Vec::new();
-            let mut dynamic_offset_count = 0;
-            let mut immutable_sampler_count = 0;
 
             for set_layout_binding in binding_iter {
                 let slb = set_layout_binding.borrow();
                 let mut content = native::DescriptorContent::from(slb.ty);
                 if slb.immutable_samplers {
                     content |= native::DescriptorContent::IMMUTABLE_SAMPLER;
+                    tmp_samplers.extend(immutable_sampler_iter
+                        .by_ref()
+                        .take(slb.count)
+                        .enumerate()
+                        .map(|(array_index, sm)| TempSampler {
+                            sampler: sm.borrow().0.clone(),
```

Comment:
Alternatively we could create the key tuple here immediately instead mapping through the temporary struct – `((binding, array_index), sampler)`

Reply:
Yes, that is an option. I don't mind either way :)